File: lessons/00-workflow-or-agent/python/src/boundrelay_m0/bounded_router.py
from dataclasses import dataclass
from typing import Mapping

from .schemas import validate_route_decision
from .types import FailureCode, RouteDecision


@dataclass(frozen=True)
class RejectedRoute:
    failure_code: FailureCode
    rejected_route: str | None


def route_decision(raw: object) -> RouteDecision | RejectedRoute:
    validated = validate_route_decision(raw)
    if validated.ok:
        return validated.value
    rejected_route: str | None = None
    if isinstance(raw, Mapping) and isinstance(raw.get("route"), str):
        rejected_route = raw["route"]
    return RejectedRoute("INVALID_ROUTE_DECISION", rejected_route)

File: lessons/00-workflow-or-agent/typescript/src/bounded-router.ts
import {validateRouteDecision} from "./schemas.js";
import type {FailureCode, RouteDecision} from "./types.js";

export type RouteDecisionResult =
  | {ok: true; decision: RouteDecision}
  | {ok: false; failureCode: FailureCode; rejectedRoute: string | null};

export function routeDecision(raw: unknown): RouteDecisionResult {
  const validated = validateRouteDecision(raw);
  if (validated.ok) {
    return {ok: true, decision: validated.value};
  }

  const rejectedRoute = typeof raw === "object"
    && raw !== null
    && "route" in raw
    && typeof (raw as Record<string, unknown>).route === "string"
    ? (raw as Record<string, unknown>).route as string
    : null;

  return {ok: false, failureCode: "INVALID_ROUTE_DECISION", rejectedRoute};
}
